### ISSUE TITLE
V4L2 decoder cannot play videos

### ISSUE THREAD

**Issue by EwanRoycroft** [bbc/chromium#3]
### Description
After building with the V4L2 decoder, videos will not play. 

When playing a normal MP4 file, the video will remain black and the timestamp will not advance. When playing a YouTube video, it shows a buffer animation indefinitely. When watching iPlayer, the video appears to have started, but it will remain black. 

Running `lsof /dev/video*` shows that Chromium has opened `/dev/video10`, and `chrome://media-internals` reports that Chromium is decoding H.264 content with `MojoVideoDecoder`. The FFmpeg demuxer displays resolution/encoding information about the streams, suggesting that the videos have been downloaded, but cannot be played.

### More information
| Attribute | Value |
| --- | --- |
| **Version** | 86.0.4184.0 (32-bit) |
| **Build** | Release |
| **Platform** | Ozone-GBM |
| **Target** | `chrome` |
| **Device** | Raspberry Pi 4B (4GB) |
| **OS** | Raspbian 10 "buster" |

### Steps to reproduce
Run Ozone-GBM with V4L2 support on the RPi 4. Try to play a H.264 video.

#### GN arguments
(Ignoring any additional debug/release/codec args.)
```
use_ozone = true
ozone_platform_x11 = false
ozone_platform_gbm = true
ozone_platform_headless = false
use_vc4_minigbm = true

ffmpeg_branding = "Chrome"
proprietary_codecs = true
use_v4l2_codec = true
v4l2_decoder_fmt_vp9 = false
v4l2_decoder_backend_stateless = false

target_os="chromeos"
target_cpu = "arm"
use_sysroot = true
target_sysroot = "//build/linux/debian_sid_arm-sysroot"

enable_nacl = false
blink_symbol_level = 0
```

#### Command
```shell
EGL_PLATFORM=surfaceless ./chrome --ozone-platform=gbm --force-system-compositor-mode [--ignore-gpu-blacklist --enable-gpu-rasterization]
```

### `media-internals` log
```json
{
  "22:0": {
    "id": "22:0",
    "properties": {
      "render_id": 22,
      "player_id": 0,
      "origin_url": "http://192.168.0.17/",
      "kFrameUrl": "http://192.168.0.17/videos/h264/bbb_1080p30.mp4",
      "kFrameTitle": "",
      "url": "http://192.168.0.17/videos/h264/bbb_1080p30.mp4",
      "kTotalBytes": 276134947,
      "kIsStreaming": false,
      "kIsSingleOrigin": true,
      "kIsRangeHeaderSupported": true,
      "pipeline_state": "kPlaying",
      "info": "Effective playback rate changed from 0 to 1",
      "debug": "Warning, FFmpegDemuxer failed to create a valid/supported audio decoder configuration from muxed stream, config:codec: unknown, profile: unknown, bytes_per_channel: 0, channel_layout: 5.1, channels: 6, samples_per_second: 48000, sample_format: Unknown sample format, bytes_per_frame: 0, seek_preroll: 0us, codec_delay: 0, has extra data: false, encryption scheme: Unencrypted, discard decoder delay: true",
      "kAudioTracks": [
        {
          "bytes per channel": 2,
          "bytes per frame": 4,
          "channel layout": "STEREO",
          "channels": 2,
          "codec": "mp3",
          "codec delay": 0,
          "discard decoder delay": true,
          "encryption scheme": "Unencrypted",
          "has extra data": false,
          "profile": "unknown",
          "sample format": "Signed 16-bit planar",
          "samples per second": 48000,
          "seek preroll": "0us"
        }
      ],
      "kVideoTracks": [
        {
          "alpha mode": "is_opaque",
          "codec": "h264",
          "coded size": "1920x1080",
          "color space": "{primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}",
          "encryption scheme": "Unencrypted",
          "has extra data": true,
          "hdr metadata": "unset",
          "natural size": "1920x1080",
          "orientation": "0°",
          "profile": "h264 high",
          "visible rect": "0,0 1920x1080"
        }
      ],
      "kMaxDuration": 634.533333,
      "kStartTime": 0,
      "kBitrate": 3481424,
      "duration": 634.533333,
      "kIsAudioDecryptingDemuxerStream": false,
      "kAudioDecoderName": "FFmpegAudioDecoder",
      "kIsPlatformAudioDecoder": false,
      "kIsVideoDecryptingDemuxerStream": false,
      "kVideoDecoderName": "MojoVideoDecoder",
      "kIsPlatformVideoDecoder": true,
      "audio_buffering_state": {
        "state": "BUFFERING_HAVE_ENOUGH"
      },
      "event": "kPlay"
    },
    "allEvents": [
      {
        "time": 0,
        "key": "origin_url",
        "value": "http://192.168.0.17/"
      },
      {
        "time": 0.05599999986588955,
        "key": "kFrameUrl",
        "value": "http://192.168.0.17/videos/h264/bbb_1080p30.mp4"
      },
      {
        "time": 0.07700000051409006,
        "key": "kFrameTitle",
        "value": ""
      },
      {
        "time": 1.0290000000968575,
        "key": "url",
        "value": "http://192.168.0.17/videos/h264/bbb_1080p30.mp4"
      },
      {
        "time": 89.34900000039488,
        "key": "kTotalBytes",
        "value": 276134947
      },
      {
        "time": 89.37200000043958,
        "key": "kIsStreaming",
        "value": false
      },
      {
        "time": 89.37899999972433,
        "key": "kIsSingleOrigin",
        "value": true
      },
      {
        "time": 89.38200000021607,
        "key": "kIsRangeHeaderSupported",
        "value": true
      },
      {
        "time": 281.1919999998063,
        "key": "pipeline_state",
        "value": "kStarting"
      },
      {
        "time": 2997.611999999732,
        "key": "info",
        "value": "FFmpegDemuxer: created video stream, config codec: h264, profile: h264 high, level: not available, alpha_mode: is_opaque, coded size: [1920,1080], visible rect: [0,0,1920,1080], natural size: [1920,1080], has extra data: true, encryption scheme: Unencrypted, rotation: 0°, flipped: 0, color space: {primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}"
      },
      {
        "time": 2997.907999999821,
        "key": "info",
        "value": "FFmpegDemuxer: created audio stream, config codec: mp3, profile: unknown, bytes_per_channel: 2, channel_layout: STEREO, channels: 2, samples_per_second: 48000, sample_format: Signed 16-bit planar, bytes_per_frame: 4, seek_preroll: 0us, codec_delay: 0, has extra data: false, encryption scheme: Unencrypted, discard decoder delay: true"
      },
      {
        "time": 2998.0590000003576,
        "key": "debug",
        "value": "Warning, FFmpegDemuxer failed to create a valid/supported audio decoder configuration from muxed stream, config:codec: unknown, profile: unknown, bytes_per_channel: 0, channel_layout: 5.1, channels: 6, samples_per_second: 48000, sample_format: Unknown sample format, bytes_per_frame: 0, seek_preroll: 0us, codec_delay: 0, has extra data: false, encryption scheme: Unencrypted, discard decoder delay: true"
      },
      {
        "time": 2998.088999999687,
        "key": "info",
        "value": "FFmpegDemuxer: skipping invalid or unsupported audio track"
      },
      {
        "time": 2998.1500000003725,
        "key": "kAudioTracks",
        "value": [
          {
            "bytes per channel": 2,
            "bytes per frame": 4,
            "channel layout": "STEREO",
            "channels": 2,
            "codec": "mp3",
            "codec delay": 0,
            "discard decoder delay": true,
            "encryption scheme": "Unencrypted",
            "has extra data": false,
            "profile": "unknown",
            "sample format": "Signed 16-bit planar",
            "samples per second": 48000,
            "seek preroll": "0us"
          }
        ]
      },
      {
        "time": 2998.230000000447,
        "key": "kVideoTracks",
        "value": [
          {
            "alpha mode": "is_opaque",
            "codec": "h264",
            "coded size": "1920x1080",
            "color space": "{primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}",
            "encryption scheme": "Unencrypted",
            "has extra data": true,
            "hdr metadata": "unset",
            "natural size": "1920x1080",
            "orientation": "0°",
            "profile": "h264 high",
            "visible rect": "0,0 1920x1080"
          }
        ]
      },
      {
        "time": 2998.354000000283,
        "key": "kMaxDuration",
        "value": 634.533333
      },
      {
        "time": 2998.3629999998957,
        "key": "kStartTime",
        "value": 0
      },
      {
        "time": 2998.3700000001118,
        "key": "kBitrate",
        "value": 3481424
      },
      {
        "time": 2998.104000000283,
        "key": "duration",
        "value": 634.533333
      },
      {
        "time": 3029.6650000000373,
        "key": "kIsAudioDecryptingDemuxerStream",
        "value": false
      },
      {
        "time": 3029.7029999997467,
        "key": "kAudioDecoderName",
        "value": "FFmpegAudioDecoder"
      },
      {
        "time": 3029.713999999687,
        "key": "kIsPlatformAudioDecoder",
        "value": false
      },
      {
        "time": 3029.888000000268,
        "key": "info",
        "value": "Selected FFmpegAudioDecoder for audio decoding, config: codec: mp3, profile: unknown, bytes_per_channel: 2, channel_layout: STEREO, channels: 2, samples_per_second: 48000, sample_format: Signed 16-bit planar, bytes_per_frame: 4, seek_preroll: 0us, codec_delay: 0, has extra data: false, encryption scheme: Unencrypted, discard decoder delay: true"
      },
      {
        "time": 3030.256000000052,
        "key": "info",
        "value": "Failed to initialize DecryptingVideoDecoder"
      },
      {
        "time": 3080.2400000002235,
        "key": "kIsVideoDecryptingDemuxerStream",
        "value": false
      },
      {
        "time": 3080.251000000164,
        "key": "kVideoDecoderName",
        "value": "MojoVideoDecoder"
      },
      {
        "time": 3080.2620000001043,
        "key": "kIsPlatformVideoDecoder",
        "value": true
      },
      {
        "time": 3080.321000000462,
        "key": "info",
        "value": "Selected MojoVideoDecoder for video decoding, config: codec: h264, profile: h264 high, level: not available, alpha_mode: is_opaque, coded size: [1920,1080], visible rect: [0,0,1920,1080], natural size: [1920,1080], has extra data: true, encryption scheme: Unencrypted, rotation: 0°, flipped: 0, color space: {primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}"
      },
      {
        "time": 3080.4890000000596,
        "key": "pipeline_state",
        "value": "kPlaying"
      },
      {
        "time": 3094.224000000395,
        "key": "audio_buffering_state",
        "value": {
          "state": "BUFFERING_HAVE_ENOUGH"
        }
      },
      {
        "time": 11318.55400000047,
        "key": "info",
        "value": "Effective playback rate changed from 0 to 1"
      },
      {
        "time": 11319.026999999769,
        "key": "event",
        "value": "kPlay"
      }
    ],
    "lastRendered": 0,
    "firstTimestamp_": 6433417.386
  }
}
```

**Comment by EwanRoycroft** [comment]
### V4L2 kernel debug
Turned on [V4L2 debug messages](https://www.raspberrypi.org/forums/viewtopic.php?t=227094) with:
```shell
echo '3' | sudo tee -a /sys/class/video4linux/video10/dev_debug
```

After running `chrome` and attempting to play an MP4 video, `dmesg` logged the following:
```
[23127.344209] videodev: v4l2_open: video10: open (0)
[23127.344280] video10: VIDIOC_ENUM_FMT
[23127.344294] video10: VIDIOC_ENUM_FMT
[23127.344305] video10: VIDIOC_ENUM_FMT
[23127.344314] video10: VIDIOC_ENUM_FMT: error -22
[23127.344339] videodev: v4l2_release: video10: release
[23127.344969] videodev: v4l2_open: video10: open (0)
[23127.345022] video10: VIDIOC_ENUM_FMT
[23127.345033] video10: VIDIOC_ENUM_FMT
[23127.345043] video10: VIDIOC_ENUM_FMT
[23127.345053] video10: VIDIOC_ENUM_FMT: error -22
[23127.345089] video10: VIDIOC_ENUM_FRAMESIZES
[23127.345100] video10: VIDIOC_QUERYCTRL: error -22
[23127.345114] video10: VIDIOC_ENUM_FRAMESIZES
[23127.345141] videodev: v4l2_release: video10: release
[23174.194771] videodev: v4l2_open: video10: open (0)
[23174.194875] video10: VIDIOC_ENUM_FMT
[23174.194897] video10: VIDIOC_ENUM_FMT
[23174.194910] video10: VIDIOC_ENUM_FMT
[23174.194922] video10: VIDIOC_ENUM_FMT: error -22
[23174.194959] videodev: v4l2_release: video10: release
[23174.203947] videodev: v4l2_open: video10: open (0)
[23174.204272] video10: VIDIOC_ENUM_FMT
[23174.204291] video10: VIDIOC_ENUM_FMT
[23174.204302] video10: VIDIOC_ENUM_FMT
[23174.204314] video10: VIDIOC_ENUM_FMT: error -22
[23174.204327] video10: VIDIOC_ENUM_FRAMESIZES
[23174.204340] video10: VIDIOC_QUERYCTRL: error -22
[23174.204355] video10: VIDIOC_ENUM_FRAMESIZES
[23174.204380] videodev: v4l2_release: video10: release
[23174.415525] videodev: v4l2_open: video10: open (0)
[23174.415795] video10: VIDIOC_ENUM_FMT
[23174.415817] video10: VIDIOC_ENUM_FMT
[23174.415828] video10: VIDIOC_ENUM_FMT
[23174.415838] video10: VIDIOC_ENUM_FMT: error -22
[23174.415867] videodev: v4l2_release: video10: release
[23174.416699] videodev: v4l2_open: video10: open (0)
[23174.416848] videodev: v4l2_open: video10: open (0)
[23174.416913] videodev: v4l2_release: video10: release
[23174.416937] video10: VIDIOC_QUERYCAP
[23174.416955] video10: VIDIOC_REQBUFS
[23174.416969] video10: VIDIOC_REQBUFS
[23174.417083] videodev: v4l2_open: video10: open (0)
[23174.417143] video10: VIDIOC_ENUM_FMT
[23174.417155] video10: VIDIOC_ENUM_FMT
[23174.417164] video10: VIDIOC_ENUM_FMT
[23174.417173] video10: VIDIOC_ENUM_FMT: error -22
[23174.417190] videodev: v4l2_release: video10: release
[23174.417845] videodev: v4l2_open: video10: open (0)
[23174.417914] video10: VIDIOC_ENUM_FMT
[23174.417927] video10: VIDIOC_ENUM_FMT
[23174.417938] video10: VIDIOC_ENUM_FMT
[23174.417947] video10: VIDIOC_ENUM_FMT: error -22
[23174.417958] video10: VIDIOC_ENUM_FRAMESIZES
[23174.417967] video10: VIDIOC_QUERYCTRL: error -22
[23174.417979] video10: VIDIOC_ENUM_FRAMESIZES
[23174.417998] videodev: v4l2_release: video10: release
[23174.418320] video10: VIDIOC_SUBSCRIBE_EVENT
[23174.418339] video10: VIDIOC_ENUM_FMT
[23174.418353] video10: VIDIOC_ENUM_FMT
[23174.418363] video10: VIDIOC_ENUM_FMT
[23174.418372] video10: VIDIOC_ENUM_FMT: error -22
[23174.418382] video10: VIDIOC_ENUM_FRAMESIZES
[23174.418395] video10: VIDIOC_S_FMT
[23174.418407] video10: VIDIOC_G_FMT
[23174.429150] video10: VIDIOC_REQBUFS
[23174.429181] video10: VIDIOC_QUERYBUF
[23174.429197] video10: VIDIOC_QUERYBUF
[23174.429209] video10: VIDIOC_QUERYBUF
[23174.429221] video10: VIDIOC_QUERYBUF
[23174.429232] video10: VIDIOC_QUERYBUF
[23174.429278] video10: VIDIOC_QUERYBUF
[23174.429291] video10: VIDIOC_QUERYBUF
[23174.429301] video10: VIDIOC_QUERYBUF
[23174.429312] video10: VIDIOC_QUERYBUF
[23174.429324] video10: VIDIOC_QUERYBUF
[23174.429335] video10: VIDIOC_QUERYBUF
[23174.429347] video10: VIDIOC_QUERYBUF
[23174.429359] video10: VIDIOC_QUERYBUF
[23174.429369] video10: VIDIOC_QUERYBUF
[23174.429403] video10: VIDIOC_QUERYBUF
[23174.429414] video10: VIDIOC_QUERYBUF
[23174.432158] video10: VIDIOC_STREAMON
[23174.434789] videodev: v4l2_mmap: video10: mmap (0)
[23174.434851] video10: VIDIOC_QBUF
[23174.435587] videodev: v4l2_mmap: video10: mmap (0)
[23174.440378] video10: VIDIOC_QBUF
[23174.443743] video10: VIDIOC_DQBUF: error -11
[23174.443810] videodev: v4l2_mmap: video10: mmap (0)
[23174.443853] video10: VIDIOC_QBUF
[23174.443902] videodev: v4l2_mmap: video10: mmap (0)
[23174.443970] video10: VIDIOC_QBUF
[23174.444029] videodev: v4l2_mmap: video10: mmap (0)
[23174.444112] video10: VIDIOC_QBUF
[23174.444168] videodev: v4l2_mmap: video10: mmap (0)
[23174.444290] video10: VIDIOC_QBUF
[23174.444354] videodev: v4l2_mmap: video10: mmap (0)
[23174.445940] video10: VIDIOC_QBUF
[23174.446006] videodev: v4l2_mmap: video10: mmap (0)
[23174.453164] video10: VIDIOC_QBUF
[23174.453258] videodev: v4l2_mmap: video10: mmap (0)
[23174.453343] video10: VIDIOC_QBUF
[23174.453391] videodev: v4l2_mmap: video10: mmap (0)
[23174.453445] video10: VIDIOC_QBUF
[23174.453485] videodev: v4l2_mmap: video10: mmap (0)
[23174.453535] video10: VIDIOC_QBUF
[23174.453577] videodev: v4l2_mmap: video10: mmap (0)
[23174.455096] video10: VIDIOC_QBUF
[23174.455161] videodev: v4l2_mmap: video10: mmap (0)
[23174.457079] video10: VIDIOC_QBUF
[23174.457150] videodev: v4l2_mmap: video10: mmap (0)
[23174.457243] video10: VIDIOC_QBUF
[23174.457323] videodev: v4l2_mmap: video10: mmap (0)
[23174.457375] video10: VIDIOC_QBUF
[23174.457428] videodev: v4l2_mmap: video10: mmap (0)
[23174.457470] video10: VIDIOC_QBUF
```

Note: `VIDIOC_DQBUF: error -11`, [`EAGAIN`](https://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html) - "[no buffer was in the outgoing queue](https://www.linuxtv.org/downloads/v4l-dvb-apis-new/userspace-api/v4l/vidioc-qbuf.html)". No buffers are dequeued after this.

**Comment by EwanRoycroft** [comment]
Fixed by #7
